### PR TITLE
add DBInitializer to Ledger to init the DB according to configuration  && modify the implementation of ledger

### PR DESCRIPTION
**Pull request by cyjseagull**
add DBInitializer to Ledger to init the DB according to configuration  && modify the implementation of ledger

### CODE REVIEW

File: libledger/LedgerInterface.h [hunk 1]
```diff
@@ -24,7 +24,7 @@
 #pragma once
 #include <initializer/ParamInterface.h>
 #include <libblockchain/BlockChainInterface.h>
-#include <libblockverifier/BlockVerifierInterface.h>
+#include <libblockverifier/BlockVerifier.h>
```
Comment (wheatli) [inline libledger/LedgerInterface.h]: I think ther upper layer should not be dependent on the derived class, such as BlockVerifer and Concensus

File: libledger/Ledger.h [hunk 2]
```diff
@@ -63,28 +65,29 @@ class Ledger : public LedgerInterface
...
-    virtual std::shared_ptr<dev::consensus::Consensus> createPBFTConsensus(
-        std::shared_ptr<dev::initializer::LedgerParamInterface> param);
+    std::shared_ptr<dev::consensus::Consensus> createPBFTConsensus();
 
 private:
+    std::shared_ptr<dev::initializer::LedgerParamInterface> m_param = nullptr;
     std::shared_ptr<dev::txpool::TxPoolInterface> m_txPool = nullptr;
     std::shared_ptr<dev::blockverifier::BlockVerifierInterface> m_blockVerifier = nullptr;
     std::shared_ptr<dev::blockchain::BlockChainInterface> m_blockChain = nullptr;
     std::shared_ptr<dev::consensus::Consensus> m_consensus = nullptr;
```
Comment (wheatli) [inline libledger/Ledger.h]: why not ConsensusInterface

File: libledger/Ledger.cpp [hunk 3]
```diff
@@ -36,70 +37,93 @@ using namespace dev::blockchain;
...
 using namespace dev::sync;
-using namespace dev::db;
 namespace dev
 {
 namespace ledger
 {
-void Ledger::initLedger(std::shared_ptr<LedgerParamInterface> param)
+void Ledger::initLedger(
+    std::unordered_map<dev::Address, dev::eth::PrecompiledContract> const& preCompile,
+    dev::blockverifier::BlockVerifier::NumberHashCallBackFunction const& pFunc)
```
Comment (wheatli) [inline libledger/Ledger.cpp]: No need to include pFunc param, we just use the method of blockchain instance to init blockverfier

File: libledger/Ledger.h (hunk 2 above)
Comment (cyjseagull) [reply libledger/Ledger.h]: Has already provide `std::shared_ptr<dev::consensus::ConsensusInterface> consensus() const` interface to obtain consensusInterface.

File: libledger/Ledger.cpp (hunk 3 above)
Comment (cyjseagull) [reply libledger/Ledger.cpp]: done, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/115/files#diff-708d2961ab288bc22804247af5e4c497R46

File: libledger/LedgerInterface.h (hunk 1 above)
Comment (cyjseagull) [reply libledger/LedgerInterface.h]: done, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/115/files#diff-33936ac6c8483534b5cbb75a700a196bR27

File: libledger/Ledger.cpp [hunk 4]
```diff
@@ -22,62 +22,127 @@
...
 {
     assert(m_blockChain);
     dev::eth::PROTOCOL_ID protocol_id = getGroupProtoclID(m_groupId, ProtocolID::TxPool);
     m_txPool = std::make_shared<dev::txpool::TxPool>(
-        m_service, m_blockChain, protocol_id, m_param->txPoolParam().txPoolLimit);
+        m_service, m_blockChain, protocol_id, m_param->mutableTxPoolParam().txPoolLimit);
 }
 
 /// init blockVerifier
-void Ledger::initBlockVerifier(BlockVerifier::NumberHashCallBackFunction const& pFunc)
```
Comment (wheatli) [inline libledger/Ledger.cpp]: leave the param and setNumberHash calling,

File: libledger/Ledger.cpp [hunk 5]
```diff
@@ -22,62 +22,127 @@
...
-    m_groupId = m_param->groupId();
+    /// init dbInitializer
+    m_dbInitializer = std::make_shared<dev::ledger::DBInitializer>(m_param);
     /// init the DB
     m_dbInitializer->initDBModules(preCompile);
     /// init blockChain
     initBlockChain();
     /// intit blockVerifier
-    initBlockVerifier(pFunc);
+    initBlockVerifier();
```
Comment (wheatli) [inline libledger/Ledger.cpp]: initBlockVerfier(m_blockchain::Number2Hash)

Comment (cyjseagull) [reply libledger/Ledger.cpp]: done, please refer to `https://github.com/FISCO-BCOS/lab-bcos/pull/115/files#diff-708d2961ab288bc22804247af5e4c497R160`

**Pull request merged**